zstd.cc

```diff
@@ -129,6 +133,86 @@ size_t zstd_processor::compress(const char* input, size_t input_len, char* outpu
     return ret;
 }
 
+template<typename Func, typename... Args>
+static size_t z_checked(std::string_view what, Func&& f, Args ...args) {
+    auto ret = f(args...);
+    if (ZSTD_isError(ret)) {
+        throw std::runtime_error(fmt::format("ZSTD {} failure: {}", what, ZSTD_getErrorName(ret)));
+    }
+    return ret;
+}
+
+size_t zstd_processor::uncompress(managed_bytes_view in, bytes_ostream& os) const {
+    z_checked("decompression", ZSTD_initDStream, _dctx.get());
+
+    size_t off = os.size_bytes();
+    ZSTD_outBuffer output = { nullptr, 0, 0 };
+
+    while (!in.empty()) {
+        auto frag = in.current_fragment();
+        auto size = frag.size();
+
+        ZSTD_inBuffer input = { frag.data(), size, 0 };
+
+        while (input.pos < input.size) {
+            if (output.pos == output.size) {
+                auto need = ZSTD_DStreamOutSize();
+                output.dst = os.write_place_holder(need);
+                output.size = need;
+                output.pos = 0;
+            }
+
+            z_checked("decompression", ZSTD_decompressStream, _dctx.get(), &output, &input);
+        }
+
+        in.remove_current();
+    }
+
+    os.remove_suffix(output.size - output.pos);
+    return os.size_bytes() - off;
+}
+
+size_t zstd_processor::compress(managed_bytes_view in, bytes_ostream& os) const {
+    z_checked("compression", ZSTD_initCStream, _cctx.get(), _compression_level);
+
+    size_t off = os.size_bytes();
+    ZSTD_outBuffer output = { nullptr, 0, 0 };
+
+    while (!in.empty()) {
+        auto frag = in.current_fragment();
+        auto size = frag.size();
+
+        ZSTD_inBuffer input = { frag.data(), size, 0 };
+        ZSTD_EndDirective mode = in.size() == size ? ZSTD_e_end : ZSTD_e_continue;
+
+        for (;;) {
+            if (output.pos == output.size) {
+                auto need = std::min(ZSTD_CStreamOutSize(), ZSTD_compressBound(input.size - input.pos));
+                output.dst = os.write_place_holder(need);
+                output.size = need;
+                output.pos = 0;
+            }
+
+            auto n = z_checked("compression", ZSTD_compressStream2, _cctx.get(), &output, &input, mode);
+
+            // if last chunk, we are finished when compress returns 0
+            if (mode == ZSTD_e_end && n == 0) {
```

Comment:
Does it mean it's possible for this `for(;;)` to spinloop at the last chunk, when we have no more data to compress but internal flushes are still ongoing and we're just waiting for them to finish and `ZSTD_compressStream2` to return 0?

Reply:
Yes. Because this is how the API usage is supposed to look (zstd docs/examples). Error codes are handled by `z_checked` -> exception.

Reply:
I hope this won't cause reactor stalls.